compiler/rustc_builtin_macros/src/deriving/smart_ptr.rs

```diff
@@ -143,31 +151,220 @@ pub fn expand_deriving_smart_ptr(
     add_impl_block(impl_generics.clone(), sym::CoerceUnsized, gen_args.clone());
 }
+
+fn is_sized_marker(path: &ast::Path) -> bool {
+    const CORE_UNSIZE: [Symbol; 3] = symbols!(core::marker::Sized);
+    const STD_UNSIZE: [Symbol; 3] = symbols!(std::marker::Sized);
+    if path.segments.len() == 3 {
+        path.segments.iter().zip(CORE_UNSIZE).all(|(segment, symbol)| segment.ident.name == symbol)
+            || path
+                .segments
+                .iter()
+                .zip(STD_UNSIZE)
+                .all(|(segment, symbol)| segment.ident.name == symbol)
+    } else {
+        *path == sym::Sized
+    }
+}
+
+struct TypeSubstitution<'a> {
+    from_name: Symbol,
+    to_ty: &'a AstTy,
+    rewritten: bool,
+}
+
+impl<'a> ast::mut_visit::MutVisitor for TypeSubstitution<'a> {
+    fn visit_ty(&mut self, ty: &mut AstTy) {
+        if let Some(name) = ty.kind.is_simple_path()
+            && name == self.from_name
+        {
+            *ty = self.to_ty.clone();
+            self.rewritten = true;
+            return;
+        }
+        match &mut ty.kind {
+            ast::TyKind::Slice(_)
+            | ast::TyKind::Array(_, _)
+            | ast::TyKind::Ptr(_)
+            | ast::TyKind::Ref(_, _)
+            | ast::TyKind::BareFn(_)
+            | ast::TyKind::Never
+            | ast::TyKind::Tup(_)
+            | ast::TyKind::AnonStruct(_, _)
+            | ast::TyKind::AnonUnion(_, _)
+            | ast::TyKind::Path(_, _)
+            | ast::TyKind::TraitObject(_, _)
+            | ast::TyKind::ImplTrait(_, _)
+            | ast::TyKind::Paren(_)
+            | ast::TyKind::Typeof(_)
+            | ast::TyKind::Infer
+            | ast::TyKind::MacCall(_)
+            | ast::TyKind::Pat(_, _) => ast::mut_visit::noop_visit_ty(ty, self),
+            ast::TyKind::ImplicitSelf
+            | ast::TyKind::CVarArgs
+            | ast::TyKind::Dummy
+            | ast::TyKind::Err(_) => {}
+        }
+    }
+
+    fn visit_param_bound(&mut self, bound: &mut GenericBound) {
+        match bound {
+            GenericBound::Trait(trait_ref, _) => {
+                if trait_ref
+                    .bound_generic_params
+                    .iter()
+                    .any(|param| param.ident.name == self.from_name)
+                {
+                    return;
+                }
+                self.visit_poly_trait_ref(trait_ref);
+            }
+
+            GenericBound::Use(args, _span) => {
+                for arg in args {
+                    self.visit_precise_capturing_arg(arg);
+                }
+            }
+            GenericBound::Outlives(_) => {}
+        }
+    }
+
+    fn visit_where_predicate(&mut self, where_predicate: &mut ast::WherePredicate) {
+        match where_predicate {
+            rustc_ast::WherePredicate::BoundPredicate(bound) => {
+                if bound.bound_generic_params.iter().any(|param| param.ident.name == self.from_name)
```

Comment:
Is this for `non_lifetime_binders`? It can't happen for lifetimes.

If so, I'll make it a hard error to have `fn foo<T>() where for<T> ...`, and you can just remove this in this PR.

Reply:
Applied